zerver/tests/test_message_fetch.py

```diff
@@ -1197,6 +1197,91 @@ def test_successful_get_messages(self) -> None:
+            "num_after": 1
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(result, 'Not logged in: API authentication or user session required',
+                               status_code=401)
+
+        post_params = {
+            "anchor": 1,
+            "num_before": 1,
+            "num_after": 1,
+            "narrow": ujson.dumps([dict(operator='is', operand="private")]),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(result, "Not logged in: API authentication or user session required",
+                               status_code=401)
+
+    def test_unauthenticated_get_messages_with_web_public(self) -> None:
+        """
+        An unauthenticated call to GET /json/messages without valid
+        parameters in the `streams:web-public` narrow returns a 401.
+        """
+        post_params: Dict[str, Union[int, str, bool]] = {
+            "anchor": 1,
+            "num_before": 1,
+            "num_after": 1,
+            # "is:private" is not a is_web_public_compatible narrow.
+            "narrow": ujson.dumps([dict(operator="streams", operand="web-public"),
+                                   dict(operator="is", operand="private")]),
+        }
+        result = self.client_get("/json/messages", dict(post_params))
+        self.assert_json_error(result, 'Not logged in: API authentication or user session required',
+                               status_code=401)
+
+    def setup_web_public_test(self) -> None:
+        """
+        Send 3 messages, one in a web-public stream, one in a non web-public stream
+        and a private message.
+        """
+        user_profile = self.example_user('iago')
+        self.login('iago')
+        web_public_stream = self.make_stream('web-public-stream', is_web_public=True)
+        non_web_public_stream = self.make_stream('non-web-public-stream')
+        self.subscribe(user_profile, web_public_stream.name)
+        self.subscribe(user_profile, non_web_public_stream.name)
+
+        self.send_stream_message(user_profile, web_public_stream.name,
+                                 content="web-public message")
+        self.send_stream_message(user_profile, non_web_public_stream.name,
+                                 content="non web-public message")
+        self.send_personal_message(user_profile, self.example_user('hamlet'),
+                                   content="private message")
+        self.logout()
+
+    def test_get_messages_with_web_public(self) -> None:
+        """
+        An unauthenticated call to GET /json/messages with valid parameters
+        including `streams:web-public` narrow returns list of messages in the
+        `web-public` streams.
+        """
+        self.setup_web_public_test()
+
+        post_params = {
+            "anchor": 10000000000000000,
+            "num_before": 1,
+            "num_after": 1,
```

Comment:
Maybe these 10 so that the test setup makes sense for the below :)

Reply:
I am not sure what last 10 message would be actually. They may or may not be web-public. Right now, I generate 3 message and test among them. I will try to add more web-public messages and test.

Reply:
Well, the point is we want to be confirming that message IDs for the messages that we created that were not web-public aren't returned.  It doesn't matter what other messages are returned; we can just write the test to ignore those.

Reply:
Yeah, I am testing all the returned messages:
```python
        for msg in messages:
            self.assertEqual(msg['content'], '<p>web-public message</p>')
            self.assertEqual(msg['flags'], ['read'])
            self.assertEqual(msg['sender_email'], sender.email)
            self.assertEqual(msg['avatar_url'], avatar_url(sender))
```